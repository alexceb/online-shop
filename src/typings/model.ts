export interface ProductOption {
  color: string
  quantity: number
  power?: number[]
  storage?: string[]
}

export interface Product {
  id: number
  name: string
  brand: string
  price: string
  available: boolean
  weight: number
  options: ProductOption[]
}

export interface ProductConfiguration {
  color: string
  power?: number
  storage?: string
}

export interface CartItem extends ProductConfiguration{
  product: Product
  quantity: number
}
