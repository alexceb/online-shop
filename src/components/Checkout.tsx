import * as React from 'react';

type CheckoutProps = {

};

export const Checkout: React.FC<CheckoutProps> = props => {

  return (
    <section className="Checkout">
      <h1>Checkout page</h1>
    </section>
  );
  
};
