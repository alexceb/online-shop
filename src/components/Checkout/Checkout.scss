@import '../../resources/vars.scss';

.checkoutContainer {
	margin-top: 2rem;

	.emptyMessage {
		text-align: center;
		margin: 0 auto;
		padding: 3rem 0;
		font-size: 1.3rem;
	}

	.totalAmount {
		margin-left: auto;
		width: 10%;
    min-width: 8rem;
    text-align: right;
    padding: 1.5rem 1rem;
	}

	.nameLabel {
		font-weight: bold;
		font-size: 1rem;
		line-height: 3rem;
		margin-left: 1rem;
		width: 40%;
	}

	.label {
		font-weight: bold;
		line-height: 3rem;
		margin: 0 0.4rem 0 1.5rem;
	}

	.checkoutList {
		list-style: none;
		margin: 0;
		padding: 0;

		.checkoutItemContainer {
			padding: 1rem 0;
			margin-top: 0.3rem;
			border: 1px solid $shop-gray;

			.checkoutItemWrapper {
				display: flex;
				flex-direction: row;
				align-items: center;
				height: 3rem;
			}

			.quantity {
				display: inline-block;
				margin-left: auto;

				& > button {
					display: inline-block;
					margin: 0 1rem;
					width: 2rem;
					height: 2rem;
					font-size: 1.3rem;
					border-radius: 0.3rem;
					cursor: pointer;
					border: none;
					color: #fff;
					box-shadow: 0 2px 2px 0 rgba(0,0,0,0.14), 0 3px 1px -2px rgba(0,0,0,0.12), 0 1px 5px 0 rgba(0,0,0,0.2);
				}

				.quantityButton {
					background-color: $shop-default-green;
				}

				.removeButton {
					background-color: $shop-gray;
					margin-left: 2rem;
				}
			}
		}
	}
}