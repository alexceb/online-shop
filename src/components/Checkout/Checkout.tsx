import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { v4 } from 'node-uuid';

import { CartItem, Product, ProductOption } from '../../typings/model';
import * as styles from './Checkout.scss';
import { useEffect } from 'react';

export interface CheckoutProps extends RouteComponentProps {
  cart: CartItem[],
  total: number,
  products: Product[],
  removeItem?: (item: CartItem) => void,
  increaseAmount?: (item: CartItem) => void,
  decreaseAmount?: (item: CartItem) => void,
  increaseAmountInStock?: (item: CartItem, amount: number) => void,
  decreaseAmountInStock?: (item: CartItem, amount: number) => void,
};

export const Checkout: React.FC<CheckoutProps> = props => {

  const { 
    cart, 
    total,
    products,
    removeItem, 
    increaseAmount, 
    decreaseAmount,
    increaseAmountInStock,
    decreaseAmountInStock,
  } = props;

  const canChangeAmount = (item: CartItem, products: Product[]): boolean => {
    const currentAmount = products
      .filter((product: Product) => {
        return product.id === item.product.id;
      })[0]
      .options.filter((option: ProductOption) => {
        return option.color === item.color;
      })[0].quantity;
    
    return currentAmount > 0;
  };
  
  const increase = (item: CartItem) => {
    const payload: CartItem = {
      ...item,
      quantity: item.quantity + 1,
    };
    if (canChangeAmount(item, products)) {
      increaseAmount(payload);
      decreaseAmountInStock(item, 1);
    }
  };

  const decrease = (item: CartItem) => {
    const payload: CartItem = {
      ...item,
      quantity: item.quantity - 1,
    };
    decreaseAmount(payload);
    increaseAmountInStock(item, 1);
  }

  const removeFromCart = (item: CartItem) => {
    removeItem(item);
    increaseAmountInStock(item, item.quantity);
  }

  return (
    <section className="Checkout">
      <div className={styles.checkoutContainer}>
        {cart && cart.length > 0 ? (
          <>
            <ul className={styles.checkoutList}>
              {cart.map((item: CartItem) => (
                <li key={v4()} className={styles.checkoutItemContainer}>
                  <div className={styles.checkoutItemWrapper}>
                    <div className={styles.nameLabel}>{`(${item.product.brand}) ${item.product.name}`}</div>
                    <div>
                      <span className={styles.label}>Color</span>
                      <span>{item.color}</span>
                    </div>
                    {item.subOption.optionType ? (
                      <div>
                        <span className={styles.label}>{item.subOption.optionType}</span>
                        <span>{item.subOption.optionValue}</span>
                      </div>
                    ) : null}
                    <div className={styles.quantity}>
                      {item.quantity !== 1 ? (
                        <button className={styles.quantityButton}
                                onClick={() => decrease(item)}>&#45;</button>
                      ) : null}
                      <span>{item.quantity}</span>
                      <button className={styles.quantityButton}
                              onClick={() => increase(item)}>&#43;</button>
                      <button className={styles.removeButton}
                              onClick={() => removeFromCart(item)}>&times;</button>
                    </div>
                  </div>
                </li>
              ))}
            </ul>
            <div className={styles.totalAmount}>
              <span className={styles.nameLabel}>Total: </span>
              <span>{`${total}$`}</span>
            </div>
          </>
        ) : (
          <div className={styles.emptyMessage}>Your cart is empty</div>
        )}
      </div>
    </section>
  );
  
};
