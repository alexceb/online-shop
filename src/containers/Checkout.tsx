import { connect } from 'react-redux';
import { AnyAction } from 'redux';
import { ThunkDispatch } from 'redux-thunk';
import { withRouter } from 'react-router-dom';

import { Checkout, CheckoutProps } from '../components/Checkout/Checkout';
import { AppState } from '../reducers';
import { CartItem } from '../typings/model';
import { removeItemFromCart, increaseAmount, decreaseAmount,  } from '../actions/cart';
import { increaseAmountInStock, decreaseAmountInStock } from '../actions/products';

const mapStateToProps = (state: AppState, ownProps: CheckoutProps) => ({
  cart: state.checkout.cart,
  total: state.checkout.total,
  products: state.products.data,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<any, any, AnyAction>) => {
  return {
    removeItem: (item: CartItem) => dispatch(removeItemFromCart(item)),
    increaseAmount: (item: CartItem) => dispatch(increaseAmount(item)),
    decreaseAmount: (item: CartItem) => dispatch(decreaseAmount(item)),
    increaseAmountInStock: (item: CartItem, amount: number) => dispatch(increaseAmountInStock(item, amount)),
    decreaseAmountInStock: (item: CartItem, amount: number) => dispatch(decreaseAmountInStock(item, amount)),
  };
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Checkout));