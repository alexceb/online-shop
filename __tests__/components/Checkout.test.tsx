import * as React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';

import { CheckoutProps } from '../../src/components/Checkout/Checkout';
import CheckoutContainer from '../../src/containers/Checkout';
import { CartItem, Product } from '../../src/typings/model';

const mockStore = configureMockStore();

describe('Checkout page component', () => {

  beforeEach(() => {
    const store = mockStore({
      cart: [],
      total: 0,
    });
  })

  describe('render()', () => {
    test('Renders <Checkout /> correctly', () => {
      const products: Product[] = [];
      const cart: CartItem[] = [];
      const total: number = 0;
      const wrapper = shallow(<CheckoutContainer cart={cart} products={products} total={total} />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});