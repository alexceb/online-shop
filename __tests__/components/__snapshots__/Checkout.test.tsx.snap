// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkout page component render() Renders <Checkout /> correctly 1`] = `
<Route>
  <Component />
</Route>
`;
